mimecast_connector.py

```diff
@@ -49,83 +46,65 @@ def __init__(self):
-        action_result.add_data(response)
-        return action_result.set_status(phantom.APP_SUCCESS, "Successfully set accessKey and secretKey")
+        headers = {"Content-Type": "application/x-www-form-urlencoded"}
+
+        data = {"client_id": self._client_id, "client_secret": self._client_secret, "grant_type": "client_credentials"}
 
-    def _get_request_headers(self, uri, action_result, expired=False):
-        if not self._access_key or not self._secret_key:
-            self._login(action_result)
-            if action_result.get_status() is False:
-                self.save_progress("Failed login with given credentials")
-                return None
-        else:
-            self.save_progress("Skipped login")
-        request_id = str(uuid.uuid4())
-        hdr_date = "{} UTC".format(datetime.datetime.utcnow().strftime("%a, %d %b %Y %H:%M:%S"))
-        # The 'hmac' library on Python v3 expects the parameters in bytes (or binary format),
-        # while the 'hmac' library on Python v2 expects the parameters in string
-        # (which are stored in binary and Unicode form on Python v2 and Python v3 respectively).
-        # UnicodeDammit.unicode_markup.encode() works for both the versions, as it converts the
-        # provided input to string on Python v2 and bytes on Python v3.
-        encoded_secret_key = UnicodeDammit(self._secret_key).unicode_markup.encode("utf-8")
-        encoded_msg = UnicodeDammit(":".join([hdr_date, request_id, uri, self._app_key])).unicode_markup.encode("utf-8")
         try:
-            hmac_sha1 = hmac.new(base64.b64decode(encoded_secret_key), encoded_msg, digestmod=hashlib.sha1).digest()
-            sig = base64.encodebytes(hmac_sha1).rstrip()
+            response = requests.post(
+                f"{self._base_url}{MIMECAST_API_PATH_OAUTH_TOKEN}",
+                headers=headers,
+                data=data,
+                verify=self.get_config().get("verify_server_cert", False),
+                timeout=DEFAULT_TIMEOUT,
+            )
+
+            if response.status_code != 200:
+                return action_result.set_status(
+                    phantom.APP_ERROR, f"Failed to get access token. Status code: {response.status_code}. Response: {response.text}"
+                )
+
+            token_data = response.json()
+            self._access_token = token_data.get("access_token")
+            expires_in = token_data.get("expires_in", 3600)  # Default 1 hour if not specified
+            self._token_expires_at = datetime.utcnow() + timedelta(seconds=expires_in)
+
+            # Save token info to state
+            try:
+                self._state["access_token"] = encryption_helper.encrypt(self._access_token, self._asset_id)
+                self._state["token_expires_at"] = self._token_expires_at.isoformat()
+                self.save_state(self._state)
+            except Exception as e:
+                self.debug_print(f"Error saving token to state: {e!s}")
```

Comment:
```suggestion
            self._state["access_token"] = encryption_helper.encrypt(self._access_token, self._asset_id)
            self._state["token_expires_at"] = self._token_expires_at.isoformat()
            try:
                self.save_state(self._state)
            except Exception as e:
                self.debug_print(f"Error saving token to state: {e!s}")
```
This is just a good practice
And I think we should be more specific on an error

Reply:
Generally, we wrap the encryption_helper operation in a try-except block because it may throw an exception—for example, if the first key (access token) is not a string or is malformed. So, I added it inside a try-except block just to be safe, Many connectors follow the same approach.

Reply:
> Many connectors follow the same approach.

That doesn't really mean they follow a good practice. Actually it's an anti-pattern. The `try:except` blocks should be as short as possible and cover as narrow cases as possible to handle specifics in the best possible way.

Reply:
I’ve committed the suggested code changes.